File: 22-generate-parentheses/generate-parentheses.cpp
class Solution {
public:
  vector<string>v;
    void solve(int i , int j , int n , string s) {
        if(i == n and j == n) {
            v.push_back(s);
            return;
        }
        if(i < n) solve(i + 1 , j , n , s + '(');
        if(j < i) solve(i  , j + 1 , n , s + ')');

    }

    vector<string> generateParenthesis(int n) {
           solve(0, 0 , n , "");
        return v;
    }
    
};

File: 22-generate-parentheses/generate-parentheses.java
class Solution {
    public void generate(List<String>res,int open,int close,int max,String s){
        if(s.length()==2*max){
             res.add(s);
             return;
        }
        if(open<max){
            generate(res,open+1,close,max,s+"(");
        }
        if(close<open){
            generate(res,open,close+1,max,s+")");

        }

    }
    public List<String> generateParenthesis(int n) {
        List<String>res=new ArrayList<>();
        generate(res,0,0,n,"");
        return res;
        
    }
}